File: apps/a-demo-be-java-vertx/src/test/java/com/ademobejavx/unit/steps/UnitReportingSteps.java
package com.ademobejavx.unit.steps;

import com.ademobejavx.unit.UnitFactory;
import com.ademobejavx.support.DirectCallService;
import com.ademobejavx.support.ScenarioState;
import com.ademobejavx.support.ServiceResponse;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.junit.jupiter.api.Assertions;

public class UnitReportingSteps {

    private final ScenarioState state;

    public UnitReportingSteps(ScenarioState state) {
        this.state = state;
    }

    private DirectCallService svc() {
        return UnitFactory.getService();
    }

    @When("^alice sends GET /api/v1/reports/pl[?]from=([^&]+)&to=([^&]+)&currency=([^&]+)$")
    public void aliceSendsGetPlReport(String from, String to, String currency) throws Exception {
        String token = state.getAccessToken();
        ServiceResponse response = svc().getPlReport(token, from, to, currency);
        state.setLastResponse(response);
    }

    @Then("the income breakdown should contain {string} with amount {string}")
    public void incomeBreakdownContains(String category, String amount) {
        checkBreakdown("incomeBreakdown", category, amount);
    }

    @Then("the expense breakdown should contain {string} with amount {string}")
    public void expenseBreakdownContains(String category, String amount) {
        checkBreakdown("expenseBreakdown", category, amount);
    }

    private void checkBreakdown(String field, String category, String amount) {
        ServiceResponse response = state.getLastResponse();
        Assertions.assertNotNull(response);
        JsonObject body = response.body();
        Assertions.assertNotNull(body);
        JsonArray breakdown = body.getJsonArray(field);
        Assertions.assertNotNull(breakdown, "Expected '" + field + "' in response");
        boolean found = false;
        for (int i = 0; i < breakdown.size(); i++) {
            JsonObject entry = breakdown.getJsonObject(i);
            if (category.equals(entry.getString("category"))) {
                Assertions.assertEquals(amount, entry.getString("total"),
                        "Expected total " + amount + " for category " + category
                                + " but got " + entry.getString("total"));
                found = true;
                break;
            }
        }
        Assertions.assertTrue(found,
                "Category '" + category + "' not found in '" + field + "'");
    }
}

File: apps/demo-be-kotlin-ktor/src/test/kotlin/com/organiclever/demoktkt/unit/steps/UnitReportingSteps.kt
package com.organiclever.demoktkt.unit.steps

import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import org.junit.jupiter.api.Assertions.assertTrue

class UnitReportingSteps {

  @When("^alice sends GET /api/v1/reports/pl\\?from=([^&]+)&to=([^&]+)&currency=([A-Z]+)$")
  fun aliceSendsGetReportsPl(from: String, to: String, currency: String) {
    val token = UnitTestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val (status, body) =
      UnitHttpHelper.get("/api/v1/reports/pl?from=$from&to=$to&currency=$currency", token)
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = body
  }

  @Then("the income breakdown should contain {string} with amount {string}")
  fun theIncomeBreakdownShouldContain(category: String, amount: String) {
    val body = UnitTestWorld.lastResponseBody
    assertTrue(
      body.contains(category) && body.contains(amount),
      "Expected income breakdown to contain '$category' with amount '$amount' in: $body",
    )
  }

  @Then("the expense breakdown should contain {string} with amount {string}")
  fun theExpenseBreakdownShouldContain(category: String, amount: String) {
    val body = UnitTestWorld.lastResponseBody
    assertTrue(
      body.contains(category) && body.contains(amount),
      "Expected expense breakdown to contain '$category' with amount '$amount' in: $body",
    )
  }
}
